rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,195 @@
+#include "libstatistics_collector/topic_statistics_collector/constants.hpp"
+#include "libstatistics_collector/topic_statistics_collector/received_message_period.hpp"
+#include "libstatistics_collector/moving_average_statistics/types.hpp"
+
+namespace
+{
+/// Return the current nanoseconds (count) since epoch
+/**
+ * For now, use hard coded time instead of a node's clock (to support sim time and playback)
+ * due to node clock lifecycle issues.
+ * @return the current nanoseconds (count) since epoch
+ */
+int64_t GetCurrentNanosecondsSinceEpoch()
+{
+  const auto now = std::chrono::system_clock::now();
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
+}
+}  // namespace
+
+namespace rclcpp
+{
+namespace topic_statistics
+{
+
+using libstatistics_collector::collector::GenerateStatisticMessage;
+using metrics_statistics_msgs::msg::MetricsMessage;
+using libstatistics_collector::moving_average_statistics::StatisticData;
+
+/**
+ * Class used to collect, measure, and publish topic statistics data. Current statistics
+ * supported for subscribers are received message age and received message period.
+ *
+ * @tparam CallbackMessageT the subscribed message type
+ */
+template<typename CallbackMessageT>
+class SubscriberTopicStatistics
+{
+  using TopicStatsCollector =
+    libstatistics_collector::topic_statistics_collector::TopicStatisticsCollector<
+    CallbackMessageT>;
+  using ReceivedMessagePeriod =
+    libstatistics_collector::topic_statistics_collector::ReceivedMessagePeriodCollector<
+    CallbackMessageT>;
+
+public:
+  /// Construct a SubcriberTopicStatistics object.
+  /**
+   * This object wraps utilities, defined in libstatistics_collector, to collect,
+   * measure, and publish topic statistics data.
+   *
+   * @param node_name the name of the node, which created this instance, in order to denote
+   * topic source
+   * @param publisher instance constructed by the node in order to publish statistics data
+   */
+  SubscriberTopicStatistics(
+    const std::string & node_name,
+    const rclcpp::Publisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr & publisher)
+  : node_name_(node_name),
+    publisher_(std::move(publisher))
+  {
+    // TODO(dbbonnie): ros-tooling/aws-roadmap/issues/226, received message age
+
+    auto received_message_period = std::make_unique<ReceivedMessagePeriod>();
+    received_message_period->Start();
+    subscriber_statistics_collectors_.emplace_back(std::move(received_message_period));
+
+    window_start_ = rclcpp::Time(GetCurrentNanosecondsSinceEpoch());
+  }
+
+  virtual ~SubscriberTopicStatistics()
+  {
+    TearDown();
+  }
+
+  /// Set the timer used to publish statistics messages.
+  /**
+   * @param measurement_timer the timer to fire the publisher, created by the node
+   */
+  void SetPublisherTimer(const rclcpp::TimerBase::SharedPtr & publisher_timer)
+  {
+    publisher_timer_ = std::move(publisher_timer);
```

Comment:
This is not good right? moving a const reference?

Reply:
This is now outdated given that the timer is now created in the new constructor.